### PR TITLE
CDVD: Fix crash when launching game during hash verification

### LINKED ISSUE #13462: [BUG]: Crash when launching a game through big picture mode while verifying the hash of an ISO file
### Describe the Bug

If you try to launch a game through big picture mode while the hash of a game is being verified, PCSX2 will crash because both the UI thread and the CPU thread are calling non-reentrant CDVD functions that modify the same global state.

### Reproduction Steps

- Go into big picture mode.
- Start verifying the hash of an ISO from the Qt game properties dialog.
- Launch a game through big picture mode.
- PCSX2 will crash.

### Expected Behavior

No crash.

### PCSX2 Revision

v2.5.258

### Operating System

Windows 11

### If Linux - Specify Distro

Linux

### Logs & Dumps

Stack trace:
```
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=<optimised out>) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=<optimised out>) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=<optimised out>, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007ffff4c4527e in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007ffff4c288ff in __GI_abort () at ./stdlib/abort.c:79
#5  0x0000555555fd0c92 in CrashHandler::CrashSignalHandler (signal=<optimised out>, siginfo=<optimised out>, ctx=0x7fffffffaa00) at /home/thomas/projects/pcsx2/common/CrashHandler.cpp:325
#6  <signal handler called>
#7  ThreadedFileReader::InternalBlockSize (this=0x0) at /home/thomas/projects/pcsx2/pcsx2/CDVD/ThreadedFileReader.h:89
#8  ThreadedFileReader::ReadSync (this=0x0, pBuffer=0x5555593b63b8 <ISOreadSector(unsigned char*, unsigned int, int)::cdbuffer+24>, sector=85690, count=1) at /home/thomas/projects/pcsx2/pcsx2/CDVD/ThreadedFileReader.cpp:294
#9  0x0000555555e4577d in ISOreadSector (tempbuffer=0x555563280280 "\255g(\262\030\027\350_\213V\201\264$\240b\377\370\312\020\247\022\t#\033&\226\215\302@\261\243X\0344\v,r\205\204\004\265Aa%\234", lsn=85690, mode=3) at /home/thomas/projects/pcsx2/pcsx2/CDVD/CDVDisoReader.cpp:332
#10 0x0000555555e3cd2e in DoCDVDreadSector (buffer=buffer@entry=0x555563280280 "\255g(\262\030\027\350_\213V\201\264$\240b\377\370\312\020\247\022\t#\033&\226\215\302@\261\243X\0344\v,r\205\204\004\265Aa%\234", lsn=lsn@entry=85690, mode=mode@entry=3)
    at /home/thomas/projects/pcsx2/pcsx2/CDVD/CDVDcommon.cpp:430
#11 0x0000555555e486f9 in IsoHasher::ComputeTrackHash (this=this@entry=0x7fffffffbdd8, track=..., callback=callback@entry=0x7fffffffbf08) at /home/thomas/projects/pcsx2/pcsx2/CDVD/IsoHasher.cpp:163
#12 0x0000555555e4855b in IsoHasher::ComputeHashes (this=0x7fffffffbdd8, callback=0x7fffffffbf08) at /home/thomas/projects/pcsx2/pcsx2/CDVD/IsoHasher.cpp:131
#13 0x00005555559b7d6a in GameSummaryWidget::onVerifyClicked (this=0x55556326dd40) at /home/thomas/projects/pcsx2/pcsx2-qt/Settings/GameSummaryWidget.cpp:288
#14 0x00007ffff55f49db in void doActivate<false>(QObject*, int, void**) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Core.so.6
#15 0x00007ffff6cbe926 in QAbstractButton::clicked(bool) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Widgets.so.6
#16 0x00007ffff6cbeb9e in QAbstractButtonPrivate::emitClicked() () from /home/thomas/projects/pcsx2/deps/lib/libQt6Widgets.so.6
#17 0x00007ffff6cc03ab in QAbstractButtonPrivate::click() () from /home/thomas/projects/pcsx2/deps/lib/libQt6Widgets.so.6
#18 0x00007ffff6cc05eb in QAbstractButton::mouseReleaseEvent(QMouseEvent*) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Widgets.so.6
#19 0x00007ffff6bf4840 in QWidget::event(QEvent*) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Widgets.so.6
#20 0x00007ffff6b90d9a in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Widgets.so.6
#21 0x00007ffff6b9a225 in QApplication::notify(QObject*, QEvent*) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Widgets.so.6
#22 0x00007ffff5590938 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Core.so.6
#23 0x00007ffff6b98e70 in QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool, bool) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Widgets.so.6
#24 0x00007ffff6c08263 in QWidgetWindow::handleMouseEvent(QMouseEvent*) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Widgets.so.6
#25 0x00007ffff6c0aec0 in QWidgetWindow::event(QEvent*) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Widgets.so.6
#26 0x00007ffff6b90d9a in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Widgets.so.6
#27 0x00007ffff5590938 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Core.so.6
#28 0x00007ffff61eea9c in QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Gui.so.6
#29 0x00007ffff625697c in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Gui.so.6
#30 0x00007ffff099cdbe in xcbSourceDispatch(_GSource*, int (*)(void*), void*) () from /home/thomas/projects/pcsx2/deps/plugins/platforms/../../lib/libQt6XcbQpa.so.6
#31 0x00007ffff45145c5 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#32 0x00007ffff4573737 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#33 0x00007ffff4513a63 in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#34 0x00007ffff58996c0 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Core.so.6
#35 0x00007ffff559cfdb in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/thomas/projects/pcsx2/deps/lib/libQt6Core.so.6
#36 0x00007ffff559886a in QCoreApplication::exec() () from /home/thomas/projects/pcsx2/deps/lib/libQt6Core.so.6
#37 0x0000555555906ee2 in main (argc=1, argv=<optimised out>) at /home/thomas/projects/pcsx2/pcsx2-qt/QtHost.cpp:2417
```

### PR DESCRIPTION
**Pull request by SternXD**
Fixes #13462 
### Description of Changes
Fixed crash when launching a game through big picture mode while verifying ISO hash.

### Rationale behind Changes
When hash verification is running on the UI thread and a user launches a game from big picture mode (CPU thread), both threads simultaneously access non-reentrant CDVD functions that modify global state, causing a race condition crash.

### Suggested Testing Steps
* Open game properties for a PS2 disc image
* Click "Verify" to start hash verification
* While verification is running, launch a game from big picture mode
* What should happen is: verification stops, popup appears with message "Hash verification was cancelled because a game was launched. Please try again when no game is running."
* Make sure no crash occurs

### Did you use AI to help find, test, or implement this issue or feature?
No.

### CODE REVIEW

**Comment by TheTechnician27** [comment]
Getting the VM state every single sector iteration for this one edge case seems inefficient and inelegant both in the code and to the user. Imo, the correct response to the issue is to 1) check once at VM startup and at VM unpause if the ComputeTrackHash is running *and* 2) check once at ComputeTrackHash startup if the VM is running.

* If the VM is running at the start of ComputeTrackHash, ask the user if they'd like to pause the game until the track hash is computed or cancel the compute.
* If ComputeTrackHash is running at the start of VM, ask the user if they'd like to delay game startup until the track hash is computed or cancel the compute.

This is assuming, of course, that we can't get let both the UI thread and the CPU thread access the critical resource (CDVD) at the same time. From a user perspective, immediately canceling the compute because they didn't magically know that we haven't properly accounted for the UI and the VM accessing the CDVD at the same time is really annoying behavior.

**Review by chaoticgd — CHANGES_REQUESTED** [review]
Doesn't seem to be working.

Edit: Accidentally just discarded my review comment. Oops.

**Comment by SternXD** [comment]
It worked on my end to stop it? (though I am working on doing what tech suggested)

File: pcsx2/CDVD/IsoHasher.cpp [hunk 1]
```diff
@@ -159,6 +161,13 @@ bool IsoHasher::ComputeTrackHash(Track& track, ProgressCallback* callback)
 		if (callback->IsCancelled())
 			return false;
 
+		// If a game is trying to launch, cancel verification to prevent CDVD access race condition
+		if (VMManager::GetState() != VMState::Shutdown)
```
Comment (chaoticgd) [inline pcsx2/CDVD/IsoHasher.cpp]: I'm still getting crashes with this after I dismiss the dialog:

```
[  188.6976] isoFile error: Block index is past the end of file! (1464254 >= 0).
[  188.6977] isoFile error: Block index is past the end of file! (1464255 >= 0).
[  188.6978] isoFile error: Block index is past the end of file! (1464256 >= 0).
[  188.6979] isoFile error: Block index is past the end of file! (1464257 >= 0).
[  188.6979] isoFile error: Block index is past the end of file! (1464258 >= 0).
[  188.6980] isoFile error: Block index is past the end of file! (1464259 >= 0).
[  188.6980] isoFile error: Block index is past the end of file! (1464260 >= 0).
[  188.6980] isoFile error: Block index is past the end of file! (1464261 >= 0).
[  188.6980] isoFile error: Block index is past the end of file! (1464262 >= 0).
[  188.6981] isoFile error: Block index is past the end of file! (1464263 >= 0).
```

I ssupect this is because the IsoHasher class calls DoCDVDclose in its destructor.

It looks like the proper way to do this for now would be to use the MainWindow::cancelGameListRefresh function since VMManager::Initialize will block until it completes.

**Comment by chaoticgd** [comment]
>     * If the VM is running at the start of ComputeTrackHash, ask the user if they'd like to pause the game until the track hash is computed or cancel the compute.
> 
>     * If ComputeTrackHash is running at the start of VM, ask the user if they'd like to delay game startup until the track hash is computed or cancel the compute.

I'm not sure it's worth it to add all this extra complexity to the VM startup process for a bodge fix like this. Ideally the CDVD functions would be rewritten to be reentrant.

**Comment by TheTechnician27** [comment]
> I'm not sure it's worth it to add all this extra complexity to the VM startup process for a bodge fix like this. Ideally the CDVD functions would be rewritten to be reentrant.

Honestly fair.

File: pcsx2-qt/Settings/GameSummaryWidget.cpp [hunk 2]
```diff
@@ -269,26 +270,50 @@ void GameSummaryWidget::populateTrackList(const GameList::Entry* entry)
...
+			QMessageBox::Ignore | QMessageBox::Cancel, QMessageBox::Cancel);
+
+		if (result == QMessageBox::Cancel)
+			return;
+
+		// User chose to pause - pause the VM
+		VMManager::SetPaused(true);
 	}
 
 	IsoHasher hasher;
```
Comment (chaoticgd) [inline pcsx2-qt/Settings/GameSummaryWidget.cpp]: Still getting the error spam in the console I was getting before, again I think because IsoHasher is calling DoCDVDclose in its destructor, which is being run when GameSummaryWidget::onVerifyClicked returns. You should put it in its own scope.

File: pcsx2/VMManager.cpp [hunk 3]
```diff
@@ -1276,6 +1277,17 @@ bool VMManager::Initialize(VMBootParameters boot_params)
...
 	Host::CancelGameListRefresh();
 
+	// Check if hash verification is in progress and coordinate with it
+	if (IsoHasher::IsComputingHash())
+	{
+		// Hash computation is in progress we can't initialize the VM yet
+		// Inform the host that we need to wait for the hash to complete
+		Host::ReportErrorAsync("Hash Verification In Progress",
+			"Hash verification is currently in progress on the selected ISO. "
+			"Please wait for it to complete before launching a game, or cancel the verification.");
```
Comment (chaoticgd) [inline pcsx2/VMManager.cpp]: Minor issue here, if you keep pressing X it will keep opening multiple copies of this dialog (not a huge deal).

Comment (chaoticgd) [reply pcsx2/VMManager.cpp]: Also be careful that you're setting the atomic at the right time.

File: pcsx2-qt/Settings/GameSummaryWidget.cpp [hunk 4]
```diff
@@ -269,92 +270,129 @@ void GameSummaryWidget::populateTrackList(const GameList::Entry* entry)
...
-	hasher.ComputeHashes(&callback);
-	if (callback.IsCancelled())
-		return;
+		if (result == QMessageBox::Cancel)
+			return;
 
-	const int hash_column = hasher.IsCD() ? 5 : 4;
-	int row = 0;
+		// User chose to pause - pause the VM
+		VMManager::SetPaused(true);
```
Comment (chaoticgd) [inline pcsx2-qt/Settings/GameSummaryWidget.cpp]: Still getting crashes with this after resuming the game, specifically when the game being played is a different game to the one being hashed. I don't think this prompting to pause/unpause the game thing is a good idea, just throw up an error and be done with it.

File: pcsx2/CDVD/IsoHasher.cpp [hunk 5]
```diff
@@ -106,6 +108,7 @@ void IsoHasher::Close()
 	if (!m_is_open)
 		return;
 
+	s_is_computing_hash = false;
```
Comment (chaoticgd) [inline pcsx2/CDVD/IsoHasher.cpp]: These atomic writes are in the wrong place. The variable should be set to true before the first CDVD call and set to false after the last CDVD call.

File: pcsx2-qt/Settings/GameSummaryWidget.cpp (hunk 4 above)
Comment (chaoticgd) [reply pcsx2-qt/Settings/GameSummaryWidget.cpp]: It appears that CDVD's internal state isn't reset when you pause/unpause.

File: pcsx2/CDVD/IsoHasher.cpp (hunk 5 above)
Comment (SternXD) [reply pcsx2/CDVD/IsoHasher.cpp]: fixed

File: pcsx2-qt/Settings/GameSummaryWidget.cpp (hunk 4 above)
Comment (SternXD) [reply pcsx2-qt/Settings/GameSummaryWidget.cpp]: yeah, got rid of this just popping an error now

File: pcsx2-qt/Settings/GameSummaryWidget.cpp [hunk 6]
```diff
@@ -269,92 +270,116 @@ void GameSummaryWidget::populateTrackList(const GameList::Entry* entry)
...
-		if (!hentry->version.empty())
+		if (callback.IsCancelled())
 		{
-			setVerifyResult(tr("Verified as %1 [%2] (Version %3).")
-								.arg(QString::fromStdString(hentry->name))
-								.arg(QString::fromStdString(hentry->serial))
-								.arg(QString::fromStdString(hentry->version)));
+			early_return = true;
+		}
+		else if (hasher.WasStoppedByGameLaunch())
```
Comment (chaoticgd) [inline pcsx2-qt/Settings/GameSummaryWidget.cpp]: Looks like this is dead code now.

Comment (SternXD) [reply pcsx2-qt/Settings/GameSummaryWidget.cpp]: whoops yeah, let me get rid of that

File: pcsx2-qt/Settings/GameSummaryWidget.cpp [hunk 7]
```diff
@@ -269,92 +270,109 @@ void GameSummaryWidget::populateTrackList(const GameList::Entry* entry)
...
+				}
+			}
 		}
 	}
-	else
-	{
-		setVerifyResult(QString::fromStdString(match_error));
-	}
+
+	if (early_return)
```
Comment (chaoticgd) [inline pcsx2-qt/Settings/GameSummaryWidget.cpp]: One last thing. This seems silly, I suspect it would make more sense to return immediately rather than setting the flag, also then you don't indent the happy path an extra time.

**Review by chaoticgd — APPROVED** [review]
~~This looks good now, although to be clear this is just a hack fix. To really fix this we'd need to rewrite the CDVD code.~~

There's probably still some really minor race conditions here if the CPU thread is inside VMManager::Initialize at just the wrong place when GameSummaryWidget::onVerifyClicked is called, but as this is just a temporary solution it will do nicely. It's much better than it was.

Edit: I should also note for other reviewers that Stern tried to do this using the existing Host::CancelGameListRefresh function although had trouble. In his defence, because the code for the hash verifier runs synchronously on the UI thread instead of on a worker thread the control flow for this gets a bit tangled.

**Comment by SternXD** [comment]
Yeah, maybe I should add a comment to the code relating to this?

**Comment by chaoticgd** [comment]
> Yeah, maybe I should add a comment to the code relating to this?

There already are comments about this.

**Comment by SternXD** [comment]
ah then yeah

**Comment by chaoticgd** [comment]
Now I think about it, given that this PR has changed so it no longer interrupts the verification, would it not make more sense just to wrap CDVD in a big lock for now?

I'll wait for someone else to comment.

File: pcsx2/CDVD/IsoHasher.cpp [hunk 8]
```diff
@@ -39,12 +41,18 @@ bool IsoHasher::Open(std::string iso_path, Error* error)
 {
 	Close();
 
+	// Set atomic flag before first CDVD call
+	s_is_computing_hash = true;
```
Comment (chaoticgd) [inline pcsx2/CDVD/IsoHasher.cpp:45]: Consider the case that someone tries to rescan all games for the Game List and tries to verify the ISO file of a hash at the same time. From a brief test I did, that seems to deadlock the UI inside CDVD. The correct way to solve this is to use a mutex. You can use std::mutex::try_lock and std::mutex::unlock instead of the atomics here. Make sure to add these calls to the other places that use CDVD too (VM initialization/shutdown and the game list scanner thread). Make sure to call lock and unlock each exactly once, test error paths.

Comment (SternXD) [reply pcsx2/CDVD/IsoHasher.cpp:45]: alright I'll just make this a draft and see what I can do in that regard I didn't really consider that at first because I was only thinking about the bug that you reported

File: pcsx2/CDVD/IsoHasher.cpp [hunk 9]
```diff
@@ -78,6 +88,7 @@ bool IsoHasher::Open(std::string iso_path, Error* error)
 		if (CDVD->getTD(track, &td) < 0 || CDVD->getTD((track == tn.etrack) ? 0 : (track + 1), &next_td) < 0)
 		{
 			Error::SetString(error, fmt::format("Failed to get track range for {}", static_cast<unsigned>(track)));
+			Close();
```
Comment (chaoticgd) [inline pcsx2/CDVD/IsoHasher.cpp:91]: Close is getting called 3 times here. Once here, once in GameSummaryWidget::onVerifyClicked and once in the IsoHasher destructor.

**Review by chaoticgd — CHANGES_REQUESTED** [review]
Found some more issues. I didn't think this through properly the first time, sorry about that.

**Comment by SternXD** [comment]
> Found some more issues. I didn't think this through properly the first time, sorry about that.

No problem better we figure it out now then right after a stable release
